### PR TITLE
mtmd, llama, ggml : Update HunyuanVL support

### PR DESCRIPTION
**Pull request by ManaEstras**
## Overview

Update support for HunyuanVL vision-language model.

This PR includes:

**ggml**
- New `ggml_interpolate_sf()` API for explicit scale factor interpolation (needed for HunyuanVL's `(H+0.1)/n_grid` position embedding scaling)
- New `GGML_SCALE_FLAG_CUSTOM_SF` flag
- Fix nearest interpolation out-of-bounds access in CPU/CUDA/Metal/SYCL/Vulkan/OpenCL backends
- Add `test_interpolate_sf` test cases (38 tests covering various modes and edge cases)

**llama**
- New `HUNYUAN_VL` architecture
- Support for `rope.scaling.alpha` and `rope_dimension_sections`
- M-RoPE support for HunyuanVL text model

**mtmd**
- New `PROJECTOR_TYPE_HUNYUANVL` projector type
- Special image token layout handling (BOI + rows with newlines + EOI)
- New `set_position_mrope_hunyuanvl()` and `mtmd_decode_use_mrope_hunyuanvl()` APIs
- Add HunyuanVL smoke test in tests.sh

**convert**
- Add `HunyuanVLVisionModel` (mmproj) and `HunyuanVLTextModel` export support

## Testing

- [x] `ctest -L main` passed
- [x] `test-backend-ops` passed (38 interpolate_sf tests on CPU and Metal)
- [x] `tools/mtmd/tests.sh` smoke test added

## Additional information

HunyuanVL uses a special position embedding interpolation that differs from standard models - it requires explicit scale factors `(H+0.1)/n_grid` instead of the simple `H/n_grid` ratio. This necessitated the new `ggml_interpolate_sf()` API.

The image token layout is also non-standard: instead of a simple `nx * ny` grid, HunyuanVL uses `BOI + (patch rows with newline separators) + EOI`, which required special handling in mtmd.

# Requirements

- I have read and agree with the [contributing guidelines](https://github.com/ggml-org/llama.cpp/blob/master/CONTRIBUTING.md)
- AI usage disclosure: YES - AI was used for code review, documentation drafting, and test case suggestions. All code was written and reviewed by human contributors.

### CODE REVIEW

File: convert_hf_to_gguf.py [hunk 1]
```diff
@@ -12020,6 +12020,123 @@ def tensor_force_quant(self, name, new_name, bid, n_dims):
...
+                new_name = "mm.post_norm." + suffix[len("after_rms."):]
+            elif suffix == "image_newline":
+                new_name = "v.image_newline"
+            elif suffix == "image_sep":
+                new_name = "v.view_seperator"
+            else:
+                # image_begin, image_end -> mm.image_begin, mm.image_end
+                new_name = "mm." + suffix
+            yield (new_name, data_torch)
+            return
```
Comment (ngxson) [inline convert_hf_to_gguf.py]: please use proper tensor mapping like all other models

File: tools/mtmd/mtmd-helper.cpp [hunk 2]
```diff
@@ -180,6 +180,60 @@ struct decode_embd_batch {
         }
     }
 
+    void set_position_mrope_hunyuanvl(llama_pos pos_0, int nx, int ny, llama_seq_id seq_id, int image_count = 0) {
```
Comment (ngxson) [inline tools/mtmd/mtmd-helper.cpp]: remove any changes in `mtmd-helper`, implement it in `mtmd_image_tokens_get_decoder_pos` instead

File: tools/mtmd/mtmd.cpp [hunk 3]
```diff
@@ -37,7 +37,8 @@ struct mtmd_image_tokens {
     uint32_t nx; // number of tokens in x direction
     uint32_t ny; // number of tokens in y direction
     bool use_mrope_pos = false; // use M-RoPE position counting (the whole image is 1 temporal position)
-    uint32_t n_tokens() const { return nx * ny; }
+    uint32_t n_tokens_total = 0;
```
Comment (ngxson) [inline tools/mtmd/mtmd.cpp]: it should ne `uint32_t n_boi`, the number of BOI tokens

please cherry-pick the logic from https://github.com/ngxson/llama.cpp/pull/100

File: tools/mtmd/mtmd.h [hunk 4]
```diff
@@ -117,6 +117,10 @@ MTMD_API bool mtmd_decode_use_non_causal(mtmd_context * ctx, const mtmd_input_ch
 // whether the current model use M-RoPE for llama_decode
 MTMD_API bool mtmd_decode_use_mrope(mtmd_context * ctx);
 
+// whether the current model uses HunyuanVL-style M-RoPE
+// (token layout differs from standard 2D grid: BOI + rows-with-newlines + EOI)
+MTMD_API bool mtmd_decode_use_mrope_hunyuanvl(mtmd_context * ctx);
```
Comment (ngxson) [inline tools/mtmd/mtmd.h]: remove this API, use `decoder_pos`, see https://github.com/ngxson/llama.cpp/pull/100 (same idea)

File: tools/mtmd/models/hunyuanocr.cpp [hunk 5]
```diff
@@ -1,11 +1,35 @@
...
+        if (ph == n_grid && pw == n_grid) {
+            pos_embd = pos_patch; // no interpolation needed
+        } else {
+            pos_patch = ggml_reshape_3d(ctx0, pos_patch, n_embd, n_grid, n_grid);
+            pos_patch = ggml_permute(ctx0, pos_patch, 2, 0, 1, 3);
+            pos_patch = ggml_cont(ctx0, pos_patch);
+            pos_patch = ggml_interpolate_sf(ctx0, pos_patch, pw, ph, n_embd, 1,
+                                            GGML_SCALE_MODE_BILINEAR,
+                                            (float)(pw + 0.1f) / n_grid,
+                                            (float)(ph + 0.1f) / n_grid);
```
Comment (ngxson) [inline tools/mtmd/models/hunyuanocr.cpp:26]: IMO the new op is quite hacky (though important note is that I'm not the one who can give the approval for a new op), it's better to simply resize the embedding on CPU, and set the resized as graph input

**Review by ngxson — COMMENTED** [review]
A general note that if you want to accelerate merging this PR:
1. Do not add a new ggml op, follow the recommendation in my comment
2. If you still have a very good reason to add a new op, you should only add CPU support in this PR

The more changes you include here, the longer it take for other reviewers to approve, and the slower it can be merged.

**Comment by ManaEstras** [comment]
Thx @ngxson let me clean up some of the code and commit. I've added another two PRs regarding the same modifications as this one. so please temporarily ignore those PRs.

**Comment by ngxson** [comment]
@ManaEstras To clarify a bit, what I mean is that let's not change anything in GGML at this time, to avoid putting too much stress for backend maintainers (I'm not a backend maintainer btw, so I can only help you on the multimodal part)

My idea is that you can either:
1. (Recommended way) to call `ggml_backend_tensor_get()` to get the tensor data, resize it inside `clip_image_batch_encode` and set the resized version as input data on the graph via `set_input_f32()`
2. Or, you can use `ggml_custom_4d` but it's not very well documented, you may loss time to debug it

An alternative method is that you can also see if you can implement the same functionality with the existing `ggml_interpolate`, plus `ggml_view` to crop the output. But I can be wrong about how it work.

**Comment by wendadawen** [comment]
@ngxson Thanks for the review comments. Here's what I've addressed:
1. Tensor mapping — Replaced manual `if-elif` remapping with standard `tensor_mapping.py`.
2. XD-RoPE — Replaced the old `M-RoPE` approach entirely with a fresh `decoder_pos` based `XD-RoPE` implementation in `mtmd-helper`. No longer uses the previous M-RoPE logic.
3. ggml_interpolate_sf — Kept CPU backend only, reverted other backend changes.

**Comment by ngxson** [comment]
@wendadawen seems like you misunderstood my comments, please refer to https://github.com/ggml-org/llama.cpp/pull/22037

Btw, there are 2 similar PRs and I don't know which one you are working on
